rust-src/id/src/types.rs

```diff
@@ -50,120 +50,19 @@ use thiserror::Error;
-            let mut buf = [0u8; ACCOUNT_ADDRESS_SIZE];
-            buf.copy_from_slice(&body);
-            Ok(AccountAddress(buf))
-        } else {
-            Err("The string does not represent a valid Concordium address.")
-        }
-    }
-}
-
-impl AsRef<[u8; 32]> for AccountAddress {
-    fn as_ref(&self) -> &[u8; 32] { &self.0 }
-}
-
-impl AsMut<[u8; 32]> for AccountAddress {
-    fn as_mut(&mut self) -> &mut [u8; 32] { &mut self.0 }
-}
-
-impl SerdeSerialize for AccountAddress {
-    fn serialize<S: Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
-        let b58_str = self.to_string();
-        ser.serialize_str(&b58_str)
-    }
-}
-
-impl<'de> SerdeDeserialize<'de> for AccountAddress {
-    fn deserialize<D: Deserializer<'de>>(des: D) -> Result<Self, D::Error> {
-        des.deserialize_str(Base58Visitor)
-    }
-}
-
-struct Base58Visitor;
-
-impl<'de> Visitor<'de> for Base58Visitor {
-    type Value = AccountAddress;
-
-    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-        write!(formatter, "A base58 string, version 1.")
-    }
-
-    fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
-        v.parse::<AccountAddress>()
-            .map_err(|_| de::Error::custom("Wrong Base58 version."))
-    }
-}
-
-impl Serial for AccountAddress {
-    #[inline]
-    fn serial<B: Buffer>(&self, x: &mut B) {
-        x.write_all(&self.0)
-            .expect("Writing to buffer should succeed.")
-    }
-}
-
-impl Deserial for AccountAddress {
-    #[inline]
-    fn deserial<R: ReadBytesExt>(source: &mut R) -> ParseResult<Self> {
-        let mut buf = [0u8; ACCOUNT_ADDRESS_SIZE];
-        source.read_exact(&mut buf)?;
-        Ok(AccountAddress(buf))
-    }
-}
-
-impl AccountAddress {
-    /// Construct account address from the registration id.
-    pub fn new<C: Curve>(reg_id: &C) -> Self {
-        let mut out = [0; ACCOUNT_ADDRESS_SIZE];
-        let hasher = Sha256::new().chain(&to_bytes(reg_id));
-        out.copy_from_slice(&hasher.finalize());
-        AccountAddress(out)
-    }
-
-    /// Check whether an address is an alias of another. Two addresses that are
-    /// aliases point to the same account.
-    pub fn is_alias_of(&self, other: &AccountAddress) -> bool { self.0[0..29] == other.0[0..29] }
-
-    /// Get the `n-th` alias of an address. There are 2^24 possible aliases.
-    /// If the counter is `>= 2^24` then this function will return [`None`].
-    pub fn get_alias(&self, counter: u32) -> Option<Self> {
-        if counter < (1 << 24) {
-            let mut data = self.0;
-            data[29..].copy_from_slice(&counter.to_be_bytes()[1..]);
-            Some(Self(data))
-        } else {
-            None
-        }
-    }
+/// Construct account address from the registration id.
+pub fn account_address_from_registration_id(reg_id: &impl Curve) -> AccountAddress {
+    let mut out = [0; ACCOUNT_ADDRESS_SIZE];
+    let hasher = Sha256::new().chain(&to_bytes(reg_id));
```

Comment:
Can't we be slightly more efficient here and serialize into `Sha256`? So
```
let mut hasher = Sha256::new();
reg_id.serial(&mut hasher);
AccountAddress(hasher.finalize().into())
```

Sha256 should implement Buffer. If it does not we can add that (there is at least implementation for Sha3).

Reply:
I have added your suggestion